mad-dog/src/helpers.js

```diff
@@ -313,16 +319,94 @@ const r6 = (withNum = false) => genRandomString(6, withNum)
 /** Delay execution for n ms */
 const delay = ms => new Promise(resolve => setTimeout(resolve, ms))
 
-/** Wrapper for custom delay time */
-const waitForIndexing = async (waitTime = 5000) => {
-  logger.info(`Pausing ${waitTime}ms for discprov indexing...`)
-  await delay(waitTime)
+const getLatestIndexedBlock = async (endpoint = DISCOVERY_NODE_ENDPOINT) => {
+  return (await axios({
+    method: 'get',
+    baseURL: endpoint,
+    url: '/health_check'
+  })).data.latest_indexed_block
+}
+
+const getLatestIndexedIpldBlock = async (endpoint = DISCOVERY_NODE_ENDPOINT) => {
+  return (await axios({
+    method: 'get',
+    baseURL: endpoint,
+    url: '/ipld_block_check'
+  })).data.data.block
 }
 
-const waitForSync = async (waitTime = 20000) => {
-  logger.info(`Pausing ${waitTime}ms for sync to occur...`)
-  await delay(waitTime)
+const monitorAllUsersSyncStatus = async ({ i, libs, executeOne }) => {
+  let primary, secondary1, secondary2, primaryClockValue, secondary1ClockValue, secondary2ClockValue
+  const userId = libs.userId
+  let synced = false
+  const startTime = Date.now()
+  while (!synced && Date.now() - startTime <= MAX_SYNC_TIMEOUT) {
+    try {
+      const replicaSetClockValues = await executeOne(i, libsWrapper => {
+        return getClockValuesFromReplicaSet(libsWrapper)
+      })
+
+      primary = replicaSetClockValues[0].endpoint
+      secondary1 = replicaSetClockValues[1].endpoint
+      secondary2 = replicaSetClockValues[2].endpoint
+      primaryClockValue = replicaSetClockValues[0].clockValue
+      secondary1ClockValue = replicaSetClockValues[1].clockValue
+      secondary2ClockValue = replicaSetClockValues[2].clockValue
+
+      logger.info(`Monitoring sync for user=${userId} | (Primary) ${primary}:${primaryClockValue} - (Secondaries) ${secondary1}:${secondary1ClockValue} - ${secondary2}:${secondary2ClockValue}`)
+
+      if (secondary1ClockValue === primaryClockValue && secondary2ClockValue && primaryClockValue) {
+        synced = true
+        logger.info(`Sync completed for user=${userId}!`)
+      }
+    } catch (e) {
+      const errorMsg = `Failed sync monitoring for user=${userId}`
+      logger.error(errorMsg)
+      console.error(e)
+      throw new Error(`${errorMsg}: ${e.message}`)
+    }
+    if (!synced) { await delay(1000) }
+  }
 }
+
+/**
+ * Wait for the discovery node to catch up to the latest block on chain up to a max
+ * indexing timeout of default 5000ms.
+ * @param {*} executeOne
+ * @param {*} libsWrapper
+ * @param {number} maxIndexingTimeout default 5000ms
+ */
+const waitForLatestBlock = async ({ executeOne, maxIndexingTimeout = MAX_INDEXING_TIMEOUT, checkIpldBlockNumber = false }) => {
```

Comment:
hmm, I dunno about this @hareeshnagaraj - that means we then have to wrap the function for every new test that uses it, yeah? Seems like this function will be called super frequently? imo it works well as a generic 'wait until DP catches up to the current block at this time this function was called'

Reply:
no i don't think we would? we'd just have `waitForLatestBlock` and `waitForLatestBlacklistBlock` and tests can use one or the other - not too opinionated on this though, i'd be fine with just consolidating to a single definition that accommodates both

Reply:
I see the argument for having explicit fns to check the IPLD block indexing and regular block indexing. I consolidated the logic because I didn't want to have 2 fns with really similar logic. also, with the addition of the defaulted param `checkIpldBlockNumber`, I think this fn is okay as it is. I have no strong opinion otherwise

Reply:
so update! I removed this fn out of mad-dog and instead made 2 explicit functions in service-commands :)